libyul/backends/evm/VariableReferenceCounter.cpp

```diff
@@ -0,0 +1,83 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * Counts the number of references to a variable.
+ */
+#include <libyul/backends/evm/VariableReferenceCounter.h>
+#include <libyul/AsmAnalysisInfo.h>
+#include <libyul/AST.h>
+
+#include <libsolutil/Visitor.h>
+
+using namespace solidity::yul;
+
+void VariableReferenceCounter::operator()(Identifier const& _identifier)
+{
+	increaseRefIfFound(_identifier.name);
+}
+
+void VariableReferenceCounter::operator()(FunctionDefinition const& _function)
+{
+	Scope* originalScope = m_scope;
+
+	yulAssert(m_info.virtualBlocks.at(&_function), "");
+	m_scope = m_info.scopes.at(m_info.virtualBlocks.at(&_function).get()).get();
+	yulAssert(m_scope, "Variable scope does not exist.");
+
+	for (auto const& v: _function.returnVariables)
+		increaseRefIfFound(v.name);
+
+	(*this)(_function.body);
```

Comment:
Note the subtle change of this not using a nested instance of ``VariableReferenceCounter``, but just descending.

Reply:
But since ``VariableReferenceCounter`` is not really stateful apart from the scope, which is reset in the Block visit and after returning from the nested visit anyways, we don't need a fresh instance.

Reply:
Or well, it is stateful now, since I made the "return value" ``m_variableReferences`` part of it - but that means we'd need to have a complex merge here, if we used a new instance, so using the same is much better.

Reply:
I think it is fine use the same instance